models/repo_permission.go

```diff
@@ -329,10 +329,23 @@ func HasAccessUnit(user *User, repo *Repository, unitType UnitType, testMode Acc
 	return hasAccessUnit(x, user, repo, unitType, testMode)
 }
 
-// canBeAssigned return true if user could be assigned to a repo
+func canBeAssignedIssues(e Engine, user *User, repo *Repository) (bool, error) {
+	return hasAccessUnit(e, user, repo, UnitTypeIssues, AccessModeWrite)
```

Comment:
I don't think this change is right. Who can be assigned for an issue? We should consider it carefully.

Reply:
@lunny What are your thoughts? I didn't think about it before, but perhaps an Issue can be assigned to someone from QA or customer service that not necessarily will need to create PRs (e.g just gather information). They will probably not even need access to the code.
A more important consideration is who can _assign_ other people to an issue, as they need to have the authority to do so. I believe we're still not making that distinction .

Reply:
Private repositories: At least read access to issues (or pull requests) should be needed so you can at least see the issue/pr. You can have access to code but not read access to issues. I don't know if it makes sense to be able to assign someone which does not have write permission in the issue/pr (can you still make comments?) 
Public repositories: Anyone has at least read access so the same applies.
For me it is OK to be less restrictive and allow assigning anyone with read access, but maybe it should be a separate PR. Should I change back to UnitTypeCode to keep existing behavior (just refactor now)?

Reply:
The current code is very complex. It is already required that assignee has write access to issue / pr. It is checked by ValidateRepoMetas in routers/repo/issue.go. So I think this code shall be removed and use existing function instead.

Reply:
Perhaps the criteria for "can be assigned" should be "write on this issue OR (read this issue AND create a PR)". Because a PR can close an issue by means of commit messages when it's merged; write access to the issue is not _really_ required. In any case, if you're assigned you should have some way of closing it or comment on it for others to close.
Just for the sake of discussion, anyone assigned should be allowed to comment on it, even if they wouldn't be able to comment on other issues (but that's not for this PR).

Reply:
I think the check of "is assignable" should match users shown in the UI? The UI list is currently only based on any write access to the repository.

Reply:
I agree they should match, either way we want that criteria to be.

Reply:
I have relaxed the required for whom is assignable to anyone with write access to code, issues or pull requests. The UI only check for any write access to repo so this is to get as close as possible to that. (I know it is still not identical as I have not considered other units than these three). 
I think we shall leave additional changes of assignable out of this PR, and continue discussing this in #4329.